### PR TITLE
fix: Check project id before starting gateway

### LINKED ISSUE #4393: Confusing message about making bucket.
```
$ mc mb mygcs mydeetestbucket
mc: <ERROR> Unable to make bucket `mygcs`. Bucket name cannot be empty
Bucket created successfully `mydeetestbucket`.

$ mc mb mygcs/mydeetestbucket
mc: <ERROR> Unable to make bucket `mygcs/mydeetestbucket`. The specified bucket is not valid.
```
Also confirmed that i have permissions to create bucket on gcs endpoint directly. Below is the error being shown on the minio server.

```
mc ls mygcs
mc: <ERROR> Unable to list folder. The specified bucket is not valid.
```
however mc cp works. Verified i have storage admin perms on gcs.

```
mc cp Downloads/videojs.imageOverlay.css mygcs/deefgsbucket
...erlay.css:  163 B / 163 B  ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓  100.00% 201 B/s 0s
```

```
ERRO[11073] Unable to fetch bucket info.                  cause=Prefix access is denied: testbucket/ source=[gateway-handlers.go:830:gatewayAPIHandlers.HeadBucketHandler()] stack=~/go/src/github.com/minio/minio/cmd/gateway-gcs.go:243:(*gcsGateway).GetBucketInfo Users/Upheus1/go/src/github.com/minio/minio/cmd/gateway-handlers.go:824:(GatewayLayer).GetBucketInfo-fm ~/go/src/github.com/minio/minio/cmd/gateway-handlers.go:829:gatewayAPIHandlers.HeadBucketHandler ~/go/src/github.com/minio/minio/cmd/gateway-router.go:115:(gatewayAPIHandlers).HeadBucketHandler-fm
ERRO[11076] Unable to fetch bucket info.                  cause=Prefix access is denied: testbucket/ source=[gateway-handlers.go:830:gatewayAPIHandlers.HeadBucketHandler()] stack=~/go/src/github.com/minio/minio/cmd/gateway-gcs.go:243:(*gcsGateway).GetBucketInfo Users/Upheus1/go/src/github.com/minio/minio/cmd/gateway-handlers.go:824:(GatewayLayer).GetBucketInfo-fm ~/go/src/github.com/minio/minio/cmd/gateway-handlers.go:829:gatewayAPIHandlers.HeadBucketHandler ~/go/src/github.com/minio/minio/cmd/gateway-router.go:115:(gatewayAPIHandlers).HeadBucketHandler-fm
ERRO[11082] Unable to fetch bucket info.                  cause=Prefix access is denied: testbucket/ source=[gateway-handlers.go:830:gatewayAPIHandlers.HeadBucketHandler()] stack=~/go/src/github.com/minio/minio/cmd/gateway-gcs.go:243:(*gcsGateway).GetBucketInfo ~/go/src/github.com/minio/minio/cmd/gateway-handlers.go:824:(GatewayLayer).GetBucketInfo-fm ~/go/src/github.com/minio/minio/cmd/gateway-handlers.go:829:gatewayAPIHandlers.HeadBucketHandler ~/go/src/github.com/minio/minio/cmd/gateway-router.go:115:(gatewayAPIHandlers).HeadBucketHandler-fm
ERRO[11135] Unable to create a bucket.                    cause=Bucket name invalid: mydeetestbucket source=[gateway-handlers.go:673:gatewayAPIHandlers.PutBucketHandler()] stack=~/go/src/github.com/minio/minio/cmd/gateway-gcs.go:233:(*gcsGateway).MakeBucketWithLocation ~/go/src/github.com/minio/minio/cmd/gateway-handlers.go:671:gatewayAPIHandlers.PutBucketHandler ~/go/src/github.com/minio/minio/cmd/gateway-router.go:113:(gatewayAPIHandlers).PutBucketHandler-fm
```
## Expected Behavior
Clearer messaging on what the result of the operation is. 

## Current Behavior
 1. Error with success message is not helpful
 2. Unclear why the bucket is not valid.
Checked that bucket was never created in gcs.

## Steps to Reproduce (for bugs)
 1. Use the commands above to create a bucket.

## Your Environment

* Version used (`minio version`):
* Environment name and version (e.g. nginx 1.9.1):
* Server type and version:
* Operating System and version (`uname -a`): Darwin macOS Sierra
* Link to your project:

### PR DESCRIPTION
**Pull request by vadmeste**
## Description
Check the format of the passed project id to show meaningful error message before passing it to GCS web services.

## Motivation and Context
Fix #4393 

## How Has This Been Tested?
go test + Manual.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: cmd/gateway-gcs.go [hunk 1]
```diff
@@ -154,6 +155,19 @@ func gcsToObjectError(err error, params ...string) error {
 	return e
 }
 
+// isValidGCSProjectId - checks if a given project id is valid or not.
+// Project IDs must start with a lowercase letter and can have lowercase
+// ASCII letters, digits or hyphens. Project IDs must be between 6 and 30 characters.
```
Comment (harshavardhana) [inline cmd/gateway-gcs.go]: Can you reference a link where is this documented?.

File: cmd/gateway-gcs.go [hunk 2]
```diff
@@ -154,6 +155,19 @@ func gcsToObjectError(err error, params ...string) error {
...
+func isValidGCSProjectID(projectID string) bool {
+	if len(projectID) < 6 || len(projectID) > 30 {
+		return false
+	}
+
+	validProjectID, err := regexp.MatchString("^[a-z][a-z0-9-]*$", projectID)
+	fatalIf(err, "Unable to parse regexp string")
```
Comment (harshavardhana) [inline cmd/gateway-gcs.go]: Don't call fatalIf inside this function.. Keep a global variale for this `regexp.MusCompile()` and then use `projectID.MatchString()`

File: cmd/gateway-gcs.go [hunk 3]
```diff
@@ -177,6 +191,10 @@ func newGCSGateway(args cli.Args) (GatewayLayer, error) {
 
 	projectID := args.First()
 
+	if !isValidGCSProjectID(projectID) {
+		fatalIf(errors.New("invalid project id"), "Unable to initialize GCS gateway")
```
Comment (harshavardhana) [inline cmd/gateway-gcs.go]: This `errors.New` can to go to gcs-errors.go

File: cmd/gateway-gcs_test.go [hunk 4]
```diff
@@ -95,3 +95,26 @@ func TestToGCSPageToken(t *testing.T) {
 	}
 
 }
+
```
Comment (harshavardhana) [inline cmd/gateway-gcs_test.go:98]: A comment for the test..

File: cmd/gateway-gcs_test.go [hunk 5]
```diff
@@ -95,3 +95,26 @@ func TestToGCSPageToken(t *testing.T) {
...
+		Valid     bool
+	}{
+		{"", false},
+		{"a", false},
+		{"Abc", false},
+		{"1bcd", false},
+		{"project-id-1-project-id-more-than-31-characters-did-you-get-it", false},
+		{"project-id-1", true},
+		{"project-id-1988832", true},
+		{"projectid1414", true},
```
Comment (harshavardhana) [inline cmd/gateway-gcs_test.go]: Add one case of 'hostnames' such as valid endpoints as well.

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by harshavardhana** [comment]
> fix: Check project id before starting gateway

There is no strict convention that you need to follow here @vadmeste - just that we need to keep commit title meaningful.

**Review by harshavardhana — APPROVED** [review]

File: cmd/gateway-gcs.go [hunk 6]
```diff
@@ -154,6 +149,21 @@ func gcsToObjectError(err error, params ...string) error {
...
+var gcsProjectIDRegex = regexp.MustCompile("^[a-z][a-z0-9-]*$")
+
+// isValidGCSProjectId - checks if a given project id is valid or not.
+// Project IDs must start with a lowercase letter and can have lowercase
+// ASCII letters, digits or hyphens. Project IDs must be between 6 and 30 characters.
+// Ref: https://cloud.google.com/resource-manager/reference/rest/v1/projects#Project (projectId section)
+func isValidGCSProjectID(projectID string) bool {
+	if len(projectID) < 6 || len(projectID) > 30 {
+		return false
+	}
```
Comment (krishnasrinivas) [inline cmd/gateway-gcs.go]: `^[a-z][a-z0-9-]{5,29}$` is better instead of the above if condition?

Comment (vadmeste) [reply cmd/gateway-gcs.go]: > ^[a-z][a-z0-9-]{5,29}$ is better instead of the above if condition?

Done

**Review by harshavardhana — APPROVED** [review]

**Review by krishnasrinivas — APPROVED** [review]

**Pull request merged**